pkg/pillar/agentlog/logobject.go

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2020 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package agentlog
+
+import (
+	"encoding/json"
+	log "github.com/sirupsen/logrus"
+)
+
+// LogObject : Holds all key values pairs to be logged later
+type LogObject struct {
+	fields map[string]interface{}
+}
+
+// NewLogObject :
+// objType -> log, action, relation etc
+// objName -> App instance name, name of file being downloaded etc
+// objUUID -> UUID of the object if present (App instance UUID) or UUID of parent object
+func NewLogObject(objType string, objName string, objUUID string) *LogObject {
```

Comment:
I see this being very useful for the name and uuid which are fixed for the lifetime of the object.

But if one wants to define an object which includes dynamically changing this in the object (e.g., always log State as "state"), is there a way to do that where the log object refers to e.g., an integer in the underlying AppInstanceStatus?

Reply:
Don't see an option that golang provides for accessing fields with given name (in a variable) from a structure.
I think it would be easier if AppInstanceState implements a method of it own. Something like below:

type AppInstanceStatus struct {
    ... many many other fields
    ...
    agentlog.LogObject
}

func (appInst *AppInstanceStatus) LogState() {
    // This will replace exiting state value if present
    status.AddField("state", appInst.State)
    ...
    ... Add some other fields as required
    ...
    status.Info("App instance life cycle event")
}

And then when required call "status.LogState()"

Reply:
This example modifies the LogObject on each call; does it replace the "state" field or add a new one for each call?
Shouldn't it clone the LogObject before adding fields?

What I was trying to figure out is whether one can create some type of log object which is dynamic in that it retrieves the state on each use. Thus instead of passing a value we pass a function which is used to get the value on each call to Info() etc.

Reply:
> This example modifies the LogObject on each call; does it replace the "state" field or add a new one for each call?
@nordmark It replaces the value "state" field every time. No duplication happens.

> Shouldn't it clone the LogObject before adding fields?
In this example it need not. But, if we have a case where we add different fields every time before we log, then we should clone and add. 
In the normal case when we log a standard set of fields, we need not clone the base object.


> 
> What I was trying to figure out is whether one can create some type of log object which is dynamic in that it retrieves the state on each use. Thus instead of passing a value we pass a function which is used to get the value on each call to Info() etc.

May be we can have an interface as below:

In logobject.go:
```
type AnyObject interface {
    GetValue(key string) interface{} 
}

type LogObject struct {
    Keys              []string
    StateObject  AnyObject
    Fields            map[string]interface{}
}

func NewLogObject(stateObject AnyObject, keys []string) *LogObject {
    return &NewLogObject{StateObject: stateObject, Keys: keys}
}

func (object *LogObject) AddKey(key string) {
    object.Keys = append(object.Keys, key)
}

func (object *LogObject) Info(args ...string) {
    for _, key := range object.Keys {
        object.AddField(key, object.StatusObject.GetValue())
    }
    log.WithFields(object.Fields).Info(args...)
}
```

Any EVE object (AppInstanceStatusf or example) will implement the "AnyObject" interface (implement GetValue) function. We initialize LogObject will a list of fields to get values for and LogObject will take care of querying for these fields when we call object.Infox function.

Reply:
If we can handle returning different types (slices, integers) and getting them rendered as json that would be good. That way we can use this for different items.